File: programmers/258709.cpp
#include <string>
#include <vector>

using namespace std;

#define M 1<<11

int n, l, chk;
int sum[M][505], win[M];
vector<int> v, answer;
vector<vector<int>> _dice;

void f1(int cur, int cnt, int val, vector<int> v) {
    
    if (cnt == n / 2) {
        for (int a : v) sum[val][a]++;
        return;
    }
    if (cur >= n) return;
    
    vector<int> nv = {};
    for (int a : v) {
        for (int b : _dice[cur]) {
            nv.push_back(a + b);
        }
    }
    f1(cur + 1, cnt + 1, val | (1 << cur), nv);
    f1(cur + 1, cnt, val, v);
}

void f2(int cur, int cnt, int val) {
    
    if (cnt == n / 2) {
        int nval = 0;
        for (int i = 0; i < n; i++) {
            if ((val & 1<<i) == 0) nval |= 1<<i;
        }
        
        for (int i = 0; i < l; i++) {
            for (int j = 0; j < i; j++) {
                win[val] += sum[val][i] * sum[nval][j];
            }
        }
        return;
    }
    if (cur >= n) return;
    
    f2(cur + 1, cnt + 1, val | 1<<cur);
    f2(cur + 1, cnt, val);
}

void f3(int cur, int cnt, int val, vector<int> v) {
    
    if (cnt == n / 2) {
        if (win[chk] < win[val]) {
            chk = val;
            answer = v;
        }
        return;
    }
    if (cur >= n) return;
    
    v.push_back(cur + 1);
    f3(cur + 1, cnt + 1, val | 1<<cur, v);
    v.erase(v.end() - 1);
    f3(cur + 1, cnt, val, v);
}

vector<int> solution(vector<vector<int>> dice) {
    
    n = dice.size();
    _dice = dice;
    l = 101 * (n / 2);
    printf("n = %d\n", n);
    
    v.push_back(0);
    f1(0, 0, 0, v);
    
    f2(0, 0, 0);
    
    v = {};
    f3(0, 0, 0, v);
    
    return answer;
}

File: programmers/258709.java
import java.util.*;
import java.util.stream.*;

class Solution {
    private static final int MAX = 101;
    
    private int n;
    private int maxWin = 0;
    private int fullMask;
    private boolean[] visited;
    private int[] answer;
    private int[][] dices;
    
    public int[] solution(int[][] dice) {
        this.n = dice.length;
        this.fullMask = (1 << n) - 1;
        this.visited = new boolean[1 << (n + 1)];
        this.answer = new int[n / 2];
        this.dices = dice;
        
        searchCombination(0, 0, 0);
        return answer;
    }
    
    // 조합 탐색: n개 중 n/2개 고르기
    private void searchCombination(int cur, int cnt, int mask) {
        if (cnt == n / 2) {
            if (visited[mask]) return;
            
            visited[mask] = true;
            visited[(~mask) & fullMask] = true;
                
            int[] score = calculateScore(mask);
            
            if (maxWin < score[0]) {
                maxWin = score[0];
                answer = extractDice(mask);
            }
            if (maxWin < score[1]) {
                maxWin = score[1];
                answer = extractDice((~mask) & fullMask);
            }
            return;
        }
        if (cur == n) return;
        
        // 선택하는 경우
        searchCombination(cur + 1, cnt + 1, mask | (1 << cur));
        // 선택하지 않는 경우
        searchCombination(cur + 1, cnt, mask);
    }
    
    // 그룹에 속하는 주사위 목록 추출
    private int[] extractDice(int mask) {
        int[] result = new int[n / 2];
        int idx = 0;
        for (int i = 0; i < n; i++) {
            if ((mask & (1 << i)) != 0) {
                result[idx++] = i + 1;
            }
        }
        return result;
    }
    
    // 두 그룹의 승리 횟수 계산
    private int[] calculateScore(int mask) {
        int maxSum = (n / 2) * 100 + 1;
        int[] sumA = new int[maxSum];
        int[] sumB = new int[maxSum];
        int[] result = { 0, 0 };
        
        countSum(0, 0, mask, sumA);
        countSum(0, 0, (~mask) & fullMask, sumB);
        
        for (int i = 0; i < maxSum; i++) {
            for (int j = 0; j < maxSum; j++) {
                if (i > j) {
                    result[0] += sumA[i] * sumB[j];
                }
                else if (i < j) {
                    result[1] += sumA[i] * sumB[j];
                }
            }
        }
        return result;
    }
    
    // 해당 그룹에서 가능한 합의 경우의 수 카운팅
    private void countSum(int cur, int sum, int mask, int[] list) {
        if (cur == n) {
            list[sum]++;
            return;
        }
        
        if ((mask & (1 << cur)) != 0) {
            // 해당 주사위가 포함되는 경우
            for (int i = 0; i < 6; i++) {
                countSum(cur + 1, sum + dices[cur][i], mask, list);
            }
        }
        else {
            // 해당 주사위가 포함되지 않는 경우
            countSum(cur + 1, sum, mask, list);
        }
    }
}